src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2810 @@
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.run();
+}
+
+TEST_F(D4NFilterFixture, GetVersionedObjectRead)
+{
+  const std::string testName = "GetVersionedObjectRead";
+ 
+  boost::asio::spawn(io, [this, &testName] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    std::string instance;
+    put_version_enabled_object(testName, instance, yield);
+    put_version_suspended_object(testName, yield);
+
+    // For version enabled object
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op_enabled(objEnabled->get_read_op());
+    EXPECT_EQ(read_op_enabled->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_enabled->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
+
+    // For version suspended object
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op_suspended(objSuspended->get_read_op());
+    EXPECT_EQ(read_op_suspended->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_suspended->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
+
+    boost::system::error_code ec;
+    request req;
+    req.push("EXISTS", TEST_BUCKET + testName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_0");
+    req.push("EXISTS", TEST_BUCKET + testName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_9");
+    req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+    req.push("HGETALL", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_9");
+    req.push("HGET", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0", "version");
+    req.push("FLUSHALL");
+
+    response< int, int, int, 
+              std::map<std::string, std::string>,
+	      std::string, boost::redis::ignore_t > resp;
+
+    conn->async_exec(req, resp, yield[ec]);
+
+    ASSERT_EQ((bool)ec, false);
+    EXPECT_EQ((int)std::get<0>(resp).value(), 1);
+    EXPECT_EQ((int)std::get<1>(resp).value(), 1);
+    EXPECT_EQ((int)std::get<2>(resp).value(), 1);
+    EXPECT_EQ(std::get<3>(resp).value().size(), 10);
+
+    std::string version = std::get<4>(resp).value();
+    std::error_code err;
+    std::string testData; 
+    std::ifstream testFile; 
+
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + instance, err), true);  
+    std::string oid = instance + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + oid, err), true);     
+    testFile.open(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + oid);
+    ASSERT_EQ(testFile.is_open(), true);
+    getline(testFile, testData);
+    EXPECT_EQ(testData, "test data");
+    testFile.close();
+
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + version, err), true);  
```

Comment:
where is the version declared and defined?

Reply:
The version is declared and defined on line 803